tests/test_latent_diffusion.py

```diff
@@ -0,0 +1,343 @@
+   if continues_training:
+       image_pipe.unet.load_state_dict(torch.load('128xflawed_249.pth')) #Where to get 128xflawed_249.pth
+       starting_epoch = 250
+       optimizer.load_state_dict(torch.load('128x_optim_flawed.pth')) #Where to get 128x_optim_flawed.pth
+
+
+   #####################
+   ##   4. TRAINING   ##
+   #####################
+   starttime = time.time()
+   logger.info("Training started at {starttime}")
+   for e in range(starting_epoch, epochs):
+      losses = []
+      for step, batch in tqdm(enumerate(train_loader), total=len(train_loader)):
+         # Sample an image from dataset and make it a three channel (RGB) image
+         # Question. Wondering if original dataset is in RGB format or just grayscale?
+         clean_images, class_labels = batch
+         clean_images = torch.unsqueeze(clean_images, 1)
+         clean_images = torch.cat((clean_images, clean_images, clean_images), dim=1)
+
+         # Move data to whatever device we are using
+         clean_images = clean_images.to(device)
+         class_labels = class_labels.to(device)
+
+         # Sample noise to add to the images
+         noise = torch.randn(clean_images.shape).to(clean_images.device)
+
+         # Sample a random timestep for each image
+         timesteps = torch.randint(
+               0,
+               image_pipe.scheduler.num_train_timesteps,
+               (batch_size,),
+               device=device,
+         ).long()
+
+         # Forward diffusion process (Add noise to the clean images according to the noise magnitude at each timestep)
+         noisy_images = image_pipe.scheduler.add_noise(clean_images, noise, timesteps)
+
+         # logger.info(f" clean_images size: {clean_images.shape}, class_labels: {class_labels}")
+         # logger.info(f" noisy_images size: {noisy_images.shape}, timesteps: {timesteps}")
+
+         # Get the model prediction for the noise
+         if add_conditioning:
+               noise_pred = image_pipe.unet(noisy_images.float(), timesteps, class_labels=class_labels, return_dict=False)[0]
+         else:
+               noise_pred = image_pipe.unet(noisy_images.float(), timesteps, return_dict=False)[0]
+
+         # Compare the predicted noise with the actual noise
+         loss = F.mse_loss(noise_pred, noise)
+
+         # Update the model parameters with the optimizer based on this loss
+         loss.backward(loss)
+         losses.append(loss.item())
+
+         # Gradient accumulation:
+         if (step + 1) % grad_accumulation_steps == 0:
+               optimizer.step()
+               optimizer.zero_grad()
+
+      # Output the average loss for a given epoch
+      average_epoch_loss = sum(losses)/len(losses)
+
+      #######################
+      ##   5. VALIDATION   ##
+      #######################
+
+      validation_losses = []
+      for step, batch in tqdm(enumerate(validation_loader), total=len(validation_loader)):
```

Comment:
All the changes from the training loop should also be added to the validation loop:
 - Encode the `clean_image` variable with the VAE
 - Change the scheduler name
 - Change the diffusion model's name